newt/builder/depgraph.go

```diff
@@ -92,58 +91,86 @@ func revdepGraph(rs *resolve.ResolveSet) (DepGraph, error) {
 		return nil, err
 	}
 
-	revGm := graphMap{}
-	for parent, children := range graph {
-		// Make sure each node exists in the reverse graph.  This step is
-		// necessary for packages that have no dependers.
-		graphMapEnsure(revGm, parent)
-
-		// Add nodes for packages with with dependers.
-		for _, child := range children {
-			rParent := child.Rpkg
-			rChild := *child
-			rChild.Rpkg = parent
-			graphMapAdd(revGm, rParent, &rChild)
+	rgraph := DepGraph{}
+	for parent, entries := range graph {
+		// Ensure parent not is present in graph, even if no one depends on it.
+		if rgraph[parent] == nil {
+			rgraph[parent] = []DepEntry{}
+		}
+
+		// Reverse the dependency relationship for each child and add to the
+		// graph.
+		for _, entry := range entries {
+			rgraph[entry.PkgName] = append(rgraph[entry.PkgName], DepEntry{
+				PkgName:     parent,
+				DepExprs:    entry.DepExprs,
+				ReqApiExprs: entry.ReqApiExprs,
+				ApiExprs:    entry.ApiExprs,
+			})
 		}
 	}
 
-	return graphMapToDepGraph(revGm), nil
+	for _, entries := range rgraph {
+		SortDepEntries(entries)
+	}
+
+	return rgraph, nil
 }
 
-func depString(dep *resolve.ResolveDep) string {
-	s := fmt.Sprintf("%s", dep.Rpkg.Lpkg.FullName())
+func depString(entry DepEntry) string {
+	s := fmt.Sprintf("%s", entry.PkgName)
 
-	var reasons []string
-	if dep.Api != "" {
-		reasons = append(reasons, fmt.Sprintf("api:%s", dep.Api))
+	type ApiPair struct {
+		api   string
+		exprs []*parse.Node
 	}
 
-	exprStr := dep.ExprString()
-	if exprStr != "" {
-		reasons = append(reasons, fmt.Sprintf("syscfg:%s", exprStr))
-	}
+	if len(entry.ReqApiExprs) > 0 {
+		var apis []string
+		for api, _ := range entry.ReqApiExprs {
+			apis = append(apis, api)
+		}
+		sort.Strings(apis)
+
+		for _, api := range apis {
+			reqes := entry.ReqApiExprs[api]
+			reqdis := reqes.Disjunction().String()
+
+			apies := entry.ApiExprs[api]
+			apidis := apies.Disjunction().String()
 
-	if len(reasons) > 0 {
-		s += fmt.Sprintf("(%s)", strings.Join(reasons, " "))
+			s += "(api:" + api
+			if reqdis != "" || apidis != "" {
+				s += ",syscfg:" + reqdis
+				if apidis != "" {
+					s += ";" + apidis
+				}
+			}
+			s += ")"
+		}
+	} else {
+		dis := entry.DepExprs.Disjunction().String()
+		if dis != "" {
+			s += "(syscfg:" + dis + ")"
+		}
 	}
 
 	return s
 }
 
 func DepGraphText(graph DepGraph) string {
-	parents := make([]*resolve.ResolvePackage, 0, len(graph))
-	for lpkg, _ := range graph {
-		parents = append(parents, lpkg)
+	parents := make([]string, 0, len(graph))
```

Comment:
This same declaration style could also be used on the code above, `var apis []string`, to get a bit of extra performance (maybe)

Reply:
Good catch, thanks.  I have fixed it.